src/audio/volume/volume_hifi5.c

```diff
@@ -0,0 +1,516 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2024 Intel Corporation. All rights reserved.
+//
+// Author: Andrula Song <andrula.song@intel.com>
+
+/**
+ * \file
+ * \brief Volume HiFi5 processing implementation without peak volume detection
+ * \authors Andrula Song <andrula.song@intel.com>
+ */
+
+#include <sof/audio/buffer.h>
+#include <sof/audio/component.h>
+#include <sof/common.h>
+#include <ipc/stream.h>
+#include <stddef.h>
+#include <stdint.h>
+
+LOG_MODULE_DECLARE(volume, CONFIG_SOF_LOG_LEVEL);
+
+#include "volume.h"
+
+#if SOF_USE_HIFI(5, VOLUME)
+
+#if (!CONFIG_COMP_PEAK_VOL)
+
+#include <xtensa/tie/xt_hifi5.h>
+
+/**
+ * \brief store volume gain 4 times for xtensa multi-way intrinsic operations.
+ * Simultaneous processing 4 samples.
+ * \param[in,out] cd Volume component private data.
+ * \param[in] channels_count Number of channels to process.
+ */
+static void vol_store_gain(struct vol_data *cd, const int channels_count)
+{
+	int32_t i;
+
+	/* using for loop instead of memcpy_s(), because for loop costs less cycles */
+	for (i = 0; i < channels_count; i++) {
+		cd->vol[i] = cd->volume[i];
+		cd->vol[i + channels_count * 1] = cd->volume[i];
+		cd->vol[i + channels_count * 2] = cd->volume[i];
+		cd->vol[i + channels_count * 3] = cd->volume[i];
```

Comment:
although it looks like it wouldn't be a single `memcpy()`, but a loop of them. So you actually mean that 4 assignments are faster than a `memcpy()`? That would be logical

Reply:
I can remove the comment to avoid it to confuse. It's most of use cases just two channels. Especially with recommended memcpy_s() there is more overhead.